File: cpp/MNN/test/test_layernorm.py
import sys
# pwd 
import os
current_working_directory = os.getcwd()
print("Current working directory:", current_working_directory)

sys.path.append("../python/")  # Adjust the path as necessary
from MNN import MLayerNorm
import numpy as np


ln = MLayerNorm(768)

input = np.random.uniform(-1, 1, (1, 5, 768)).astype(np.float32)

output = ln.forward(input)

print(ln.gamma.shape, ln.beta.shape)
print(input.shape, output.shape)

save_path = "build/test/temp"
ln.gamma.tofile(os.path.join(save_path, "gamma.bin"))
ln.beta.tofile(os.path.join(save_path, "beta.bin"))
input.tofile(os.path.join(save_path, "input.bin"))
output.tofile(os.path.join(save_path, "output.bin"))

File: cpp/MNN/test/test_layernorm.cpp
#include "MNorm.hpp"
#include "test_common.hpp"

using namespace mnn;
using namespace mtb;

#define FP_T float

void test_MNorm(){
    START_TEST();
    auto gamma = load_data<FP_T>(
      "build/test/temp/gamma.bin", 1 * 768);
    auto beta = load_data<FP_T>(
      "build/test/temp/beta.bin", 1 * 768);
    auto input = load_data<FP_T>(
      "build/test/temp/input.bin", 1 * 5 * 768);
    auto output = load_data<FP_T>(
      "build/test/temp/output.bin", 1 * 5 * 768);
    
    MLayerNorm<FP_T> norm(768, 1e-5f);
    norm.fill_gamma(gamma.data(), 1 * 768);
    norm.fill_beta(beta.data(), 1 * 768);

    Tensor<FP_T> input_tensor({1, 5, 768}, input);

    Tensor<FP_T> result_tensor = norm.forward(input_tensor);

    compare_data<FP_T>(result_tensor.data().get(), 
                        output.data(), 1 * 5 * 768, 1e-5f);

    PASSLOG();
}

int main(int argc, char** argv) {
    test_MNorm();
    return 0;
}